### PR TITLE
Add `test` option for additional checks

### PR DESCRIPTION
**Pull request by hanneskuettner**
### Initial checklist

* [x] I read the support docs 
* [x] I read the contributing guide 
* [x] I agree to follow the code of conduct 
* [x] I searched issues and couldn’t find anything (or linked relevant results below) 
* [x] If applicable, I’ve added docs and tests

### Description of changes

Add a `test` option that is converted using `hast-util-is-element` and adds additional matching capabilities to exclude some external link elements as discussed in #3.

### CODE REVIEW

**Comment by wooorm** [comment]
Can you move the `is = convert()` bit up, so that it is called once: e.g., directly inside `rehypeExternalLinks`. Now it’s doing unneeded work.

I think `is` can be called in the `node.tagName === 'a'` branch? That way, if there is a test that says the element shouldn’t be converted, we don’t need to do extra work of checking protocols.

Lastly, would it be an idea to default to `'a'` as a test?
That way, people could allow `['a', 'area']`. Or include custom elements (weird but ok)?
Not sure if useful. What do you think?

**Comment by hanneskuettner** [comment]
> Can you move the `is = convert()` bit up, so that it is called once: e.g., directly inside `rehypeExternalLinks`. Now it’s doing unneeded work.

Good point!

> I think `is` can be called in the `node.tagName === 'a'` branch? That way, if there is a test that says the element shouldn’t be converted, we don’t need to do extra work of checking protocols.

Another good point, I'll also do that!

> Lastly, would it be an idea to default to `'a'` as a test?
> That way, people could allow `['a', 'area']`. Or include custom elements (weird but ok)?
> Not sure if useful. What do you think?

I thought about that as well. This could maybe lead to some unexpected behavior, since the test would then have to reimplement checking for an a tag in the simple case of just wanting to exclude some domains for example.
But maybe that is okay with an additional note in the docs?
It would certainly make this plugin more powerful.

**Comment by wooorm** [comment]
The only real benefit would be that `area` elements are supported.

Potentially custom elements too, but then it gets weird that `rel` and `target` are applied with the same semantics as `a`, because custom elements probably work differently.

Perhaps `area` could also just be added extra here for everyone: `node.tagName === 'a' || node.tagName === 'area'`. Later.

I think I’m fine with the current state!

File: test.js [hunk 1]
```diff
@@ -276,6 +276,28 @@ test('rehypeExternalLinks', async (t) => {
...
+    ),
+    '<a href="http://example.com" rel="nofollow">http</a>',
+    'should add rel to a link that matches the test function'
+  )
+
+  t.equal(
+    String(
+      await rehype()
+        .use({settings: {fragment: true}})
+        .use(rehypeExternalLinks, {test: (node) => node.properties.href === 'http://foobar.com'})
```
Comment (wooorm) [inline test.js:294]: Please change these checks from `node.properties.href` -> `node.properties && node.properties.href`. See CI.

And then run `npm test` to format everything!

Comment (hanneskuettner) [reply test.js]: `npm test` fails due to missing coverage (as does the main branch for me), but I've formatted it with `npm format` 👍

**Review by wooorm — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -36,6 +36,7 @@
   "dependencies": {
     "@types/hast": "^2.0.0",
     "extend": "^3.0.0",
+    "hast-util-is-element": "^2.1.3",
```
Comment (hanneskuettner) [inline package.json:39]: 30 seconds to late, but I just noticed that all other deps are listed in their base major version. Should this also be "^2.0.0"?

Comment (hanneskuettner) [reply package.json:39]: Perfect, that's what I figured. Didn't want it to be missed. 👍

**Comment by hanneskuettner** [comment]
Thanks for your fast handling. That was a joy!

**Pull request merged**